### PR TITLE
feat: Add `keep_at_rules` option

### PR DESCRIPTION
**Pull request by kamilzych**
Rel https://github.com/Stranger6667/css-inline/issues/265#issuecomment-2994102601

### CODE REVIEW

File: css-inline/src/html/serializer.rs [hunk 1]
```diff
@@ -114,6 +127,22 @@ impl<'a> Sink<'a> {
 
                 serializer.start_elem(&element.name, &element.attributes, style_node_id)?;
 
+                // TODO this part is the one that I don't like the most
```
Comment (Stranger6667) [inline css-inline/src/html/serializer.rs:130]: I am not 100% sure about this one, even though it is a sound approach.

The first thought that I have is that the least surprising thing would be to keep `@` rules in their original places, instead of pushing everything into a single node. This won't work for external stylesheets, so we need to store them somewhere anyway.

On the other hand, this approach is better for performance, and probably it would not be super surprising for the user anyway if we put everything under one `<style>` tag.

So, in the end, I am leaning towards this approach.

A few other notes:

1. We can reduce the performance hit of checking each node name by having two versions of `serialize` - the first one will work as before this PR. The new one will check for `head` and once it is found, it will switch to the other version of `serialize` that won't check the tag name. At the top level we check if `at_rules` are present and then call one of these two versions - I think it should minimize the impact as `head` is usually in the very beginning of the nodes vector.
2. Are there cases when `head` is not present? Maybe in HTML fragments? see `inline_fragment`

**Review by Stranger6667 — CHANGES_REQUESTED** [review]
Great start! Definitely on the right track :) I left a few comments

File: css-inline/src/html/serializer.rs (hunk 1 above)
Comment (Stranger6667) [reply css-inline/src/html/serializer.rs:130]: btw, if the user chose to keep `style` tags - I think we can only extract `@` rule from `link`, as the old `style` tags will be around anyway

**Comment by Stranger6667** [comment]
Hi @kamilzych 

Let me know if you need any assistance or if you plan to continue working on this PR. I'd be happy to help or take over in case if you don't plan to continue this PR.

**Comment by kamilzych** [comment]
@Stranger6667 Hi! Apologies for lack of activity, I simply had no time, life happened `¯\_(ツ)_/¯`

I've just pushed a commit which addresses some of your comments. I'd love to finish the whole PR but I'm not sure when I have enough time to move this forward.
That said, if you have time to take it over, I'm absolutely fine with it. If you don't, I'll do my best to finish it ASAP and in such case I'd appreciate a review of the mentioned commit and perhaps some guidance about the next steps 🙏

File: css-inline/src/html/attributes.rs [hunk 2]
```diff
@@ -123,7 +123,8 @@ pub(crate) fn hash_class_name(name: &[u8]) -> u64 {
 }
 
 /// A collection of HTML attributes.
-#[derive(Debug)]
+// TODO there must be a way for not having to instantiate Attributes class, see serializer.rs
+#[derive(Debug, Default)]
```
[the remark this answers is not in the document]
Comment (kamilzych) [reply css-inline/src/html/attributes.rs]: Addressed in https://github.com/Stranger6667/css-inline/pull/485/commits/c4adfc7da4975f8ecc8a58c50a0bd1fe594a3bbf
A the beginning I wanted to use `write_text()` but it was escaping `<` and `>` in `<style>...</style>` due to the logic it has. Therefore, I made a new `write_at_rules_style()`

File: css-inline/src/html/serializer.rs [hunk 3]
```diff
@@ -15,13 +15,17 @@ pub(crate) fn serialize_to<W: Write>(
     styles: DocumentStyleMap<'_>,
     keep_style_tags: bool,
     keep_link_tags: bool,
+    keep_at_rules: bool,
+    at_rules: Option<String>,
```
[the remark this answers is not in the document]
Comment (kamilzych) [reply css-inline/src/html/serializer.rs]: Addressed in https://github.com/Stranger6667/css-inline/pull/485/commits/c4adfc7da4975f8ecc8a58c50a0bd1fe594a3bbf
After removing `keep_at_rules: bool` flag we're back to 6 args :)

File: css-inline/src/html/serializer.rs [hunk 4]
```diff
@@ -60,6 +71,8 @@ impl<'a> Sink<'a> {
             node,
             self.keep_style_tags,
             self.keep_link_tags,
+            self.keep_at_rules,
+            self.at_rules.clone(),
```
[the remark this answers is not in the document]
Comment (kamilzych) [reply css-inline/src/html/serializer.rs]: Addressed in https://github.com/Stranger6667/css-inline/pull/485/commits/c4adfc7da4975f8ecc8a58c50a0bd1fe594a3bbf
Not sure if that's the right approach but no more `clone` call

File: css-inline/src/html/serializer.rs (hunk 1 above)
Comment (kamilzych) [reply css-inline/src/html/serializer.rs]: That's something I had no time to look closer into yet

File: css-inline/src/parser.rs [hunk 5]
```diff
@@ -98,3 +98,81 @@ impl<'i> cssparser::QualifiedRuleParser<'i> for CSSDeclarationListParser {
...
+        _: &ParserState,
+        input: &mut cssparser::Parser<'i, 't>,
+    ) -> Result<Self::QualifiedRule, cssparser::ParseError<'i, Self::Error>> {
+        let mut parser = CSSDeclarationListParser;
+        let parser = cssparser::RuleBodyParser::new(input, &mut parser);
+        let start = self.declarations.len();
+        for item in parser.flatten() {
+            self.declarations.push(item);
+        }
+        Ok((prelude, (start, self.declarations.len())))
```
[the remark this answers is not in the document]
Comment (kamilzych) [reply css-inline/src/parser.rs]: Addressed in https://github.com/Stranger6667/css-inline/pull/485/commits/c4adfc7da4975f8ecc8a58c50a0bd1fe594a3bbf

**Comment by Stranger6667** [comment]
Awesome! Thank you so much! :) I’ll be on vacation till the end of next week, but I’ll review the changes in any event :) Let’s see how it goes

File: css-inline/src/html/serializer.rs (hunk 1 above)
Comment (Stranger6667) [reply css-inline/src/html/serializer.rs]: Given the current benchmark results, the performance hit is negligible, so we can skip this. For fragments, likely we can't really do it, as there is no regular HTML structure generally. Maybe we can return an error when the user sets an option to keep at rules, but uses `inline_fragment`?

**Comment by Stranger6667** [comment]
Amazing to see that we already have a working version! There are a few minor things left; there is no time pressure, so I'm just noting what is left to do, and maybe I'll push some changes to it in the next few evenings:

- [ ] Changelog entries for the main crate + bindings
- [ ] Update all readmes with the new option + usage example
- [ ] JS bindings
- [ ] Java bindings
- [ ] Python bindings
- [ ] C bindings
- [ ] Ruby bindings
- [ ] Error for `inline_fragment` + `keep_at_rules`?

**Comment by Stranger6667** [comment]
As for the next steps, I think most of them are about adding new config option to bindings, so they compile. This flag is the same as others in terms of behavior, so it should be more or less straightforward.

about the error, we can add an extra check to inline_fragment

File: css-inline/src/main.rs [hunk 6]
```diff
@@ -199,6 +201,9 @@ OPTIONS:
 
     --keep-link-tags
         Keep "link" tags after inlining.
+
```
Comment (Stranger6667) [inline css-inline/src/main.rs:204]: There is a small place above where this flag should be added too. The handle_boolean_flag function

**Comment by kamilzych** [comment]
@Stranger6667 Hi again! I should have some time tomorrow to work on this again. Updating CHANGELOG, README and other docs should be easy. I can also take a look into 

>Error for inline_fragment + keep_at_rules?

and perhaps add a test(s) for it and other cases.

The only question I have right now is about bindings. Do you have any docs, guidance, or previous PR that could show me the ropes?

**Comment by Stranger6667** [comment]
@kamilzych 

Hi! I’ll have access to my laptop in a few hours and will provide you with more details :)

**Comment by Stranger6667** [comment]
For bindings, all the TODO entries are similar. The core of them is:

- Fill the options structure with the missing `keep_at_rules`
- Extract the `keep_at_rules` value the same way as it goes for other boolean flags like `keep_style_tags`
- Update language-specific annotations / docstrings

Depending on the language, it requires a bit different annotations, but you can just copy them from `keep_style_tags`.

For Ruby, extend the `get_kwargs` call by adding the field name to the array of strings + adding `Option<bool>` to one of the generics so the name of the new kwarg corresponds to that new `Option<bool>` position.

For Python, you need to add `keep_at_rules = False` to the `signature` macro + update the docstring the same way, so the positions match.

For C, you need to extend `CssInlinerOptions` with one new boolean and that is it.

For Java, extract a new boolean from the environment:

```rust
    let keep_at_rules = env.get_bool_field(&cfg, "keepAtRules")?;
```

And add it to the options builder.

For JavaScript, extend `options::Options` with this new boolean field, then the deserialization logic will be automatically derived. Also, you'll need to update the `InlineOptions` interface inside the `INLINE` constant. Then running `npm run build:wasm` && `npm run build:wasm-web` inside `bindings/javascript` should be enough to regenerate everything needed.

Let me know if I missed anything, will be happy to elaborate :)

**Comment by Stranger6667** [comment]
One minor detail about Ruby: bindings use the latest release from crates.io to avoid packaging issues. It is suboptimal, but I am fine with having this PR point to the local crate, so we see everything passes. Then I'll update it before the release, and maybe add a better solution to avoid breaking releases for Ruby bindings.

File: css-inline/tests/test_inlining.rs [hunk 7]
```diff
@@ -940,6 +945,27 @@ fn keep_link_tags() {
...
+@media (max-width: 600px) { h1 { font-size: 18px; } }
+p { margin: 10px; }
+</style>
+</head>
+<body>
+<h1>Hello</h1><p>World</p>
+</body>
+</html>"#;
+    let inlined = inliner.inline(html).unwrap();
+    let expected = "<html><head><style>@media (max-width: 600px)  { h1 { font-size: 18px; } }</style>\n\n</head>\n<body>\n<h1 style=\"color: blue;\">Hello</h1><p style=\"margin: 10px;\">World</p>\n\n</body></html>";
```
Comment (Stranger6667) [inline css-inline/tests/test_inlining.rs:965]: I see there is a double space - `(max-width: 600px)  {`. I guess we can keep what is in the original string without adding extra space - it should be a bit cleaner + a bit faster

**Comment by kamilzych** [comment]
@Stranger6667 Regarding error when inlining fragments. I've just checked and it seems that `keep_style_tags=true` doesn't throw any error when inlining fragment. If I understand correctly, it is pretty much the same case, right?
If so, perhaps it's better to stick with this approach for unified experience and simply ignore `keep_at_rules` for now when inlining fragments? Changing this to throw errors for both flags could be a separate issue. What do you think?

_here's the test I used for testing_
```rust
#[test]
fn inline_fragment_keep_style_tags() {
    let inliner = CSSInliner::options().keep_style_tags(true).build();
    const HTML: &str = r#"<main>
<h1>Hello</h1>
<section>
<p>who am i</p>
</section>
</main>"#;

    const CSS: &str = r#"
p {
    color: red;
}

h1 {
    color: blue;
}
"#;
    let inlined = inliner.inline_fragment(HTML, CSS).unwrap();
    assert_eq!(inlined, "<main>\n<h1 style=\"color: blue;\">Hello</h1>\n<section>\n<p style=\"color: red;\">who am i</p>\n</section>\n</main>");
}
```

**Comment by Stranger6667** [comment]
@kamilzych I completely agree with you! lets handle that part separately

**Comment by Stranger6667** [comment]
Awesome! The last failure with Ruby bindings could be resolved by running `cargo generate-lockfile` in `bindings/ruby` + copying the `Cargo.lock` file to `ext/css_inline`

**Comment by kamilzych** [comment]
> Awesome! The last failure with Ruby bindings could be resolved by running `cargo generate-lockfile` in `bindings/ruby` + copying the `Cargo.lock` file to `ext/css_inline`

Unfortunately, this is where I failed 🥲

After changing `bindings/ruby/ext/css_inline/Cargo.toml` to this
```
[dependencies.css-inline]
path = "../../../../css-inline"
version = "*"
```
running `cargo generate-lockfile` fails with `package collision in the lockfile:`

If I keep version `0.16` the command generates Lockfile, but then it will still use 0.16 from crates.io (at least that's what I think :))

**Comment by Stranger6667** [comment]
No worries! I will take a look :) I think at this point everything should be resolved and the feature is ready :) thank you so much for working on it! I plan to merge it if there are no unresolved issues and will sort out the ruby part later

**Comment by kamilzych** [comment]
Thank you too for all the patience, guidance and help 🙏 I had a lot of fun when working on it!

In the meantime, I created https://github.com/Stranger6667/css-inline/issues/514 

I will also take a look at unit tests, might add more.

**Comment by kamilzych** [comment]
@Stranger6667 I can already see some issues. Some are simply incorrect behaviour which I will try to fix, but for one I have a question. Let's consider such test:

```rust
    let html = html!("@media (max-width: 767px) { padding: 0;} h1 {background-color: blue;}", "<h1>Hello world!</h1>");
    let options = InlineOptions {
        inline_style_tags: false,
        keep_style_tags: false,
        keep_at_rules: true,
        ..Default::default()
    };
    let inliner = CSSInliner::new(options);
    let result = inliner.inline(&html).unwrap();
    assert_eq!(
        result,
        "???"
    )
```

What should we expect for such options setup? Right now, the result is `<html><head></head><body><h1>Hello world!</h1></body></html>` so `keep_at_rules` is ignored.

**Comment by Stranger6667** [comment]
> Thank you too for all the patience, guidance and help 🙏 I had a lot of fun when working on it!

I am glad to hear it! :)

Re: config options combo. From a use case point of view, I think the user might want to remove everything except at-rules, so I'd expect a `style` element with only the `@media` part. 

As `keep_style_tags` defaults to `false`, it is equivalent to `options().inline_style_tags(false).keep_at_rules(true)`. In this scenario, I read it like "don't inline anything, but keep @-rules" + the default behavior is to remove styles, so it kinda checks out. I see where the contradiction is, but I am not sure whether it would be surprising for the user if such a combo will remove all except @-rules. What do you think?


P.S., adding an unused private field could be a good idea to avoid a public API to construct this without `InlineOptions`

**Comment by kamilzych** [comment]
Yeah, with `keep_style_tags: false && keep_at_rules: true` it does work as you described and I think that setting `inline_style_tags` to false shouldn't change anything. I'll fix it.

>adding an unused private field could be a good idea to avoid a public API to construct this without InlineOptions

Not sure if I follow, where this private field should be added? 🤔

**Comment by Stranger6667** [comment]
Thanks! :)

I think adding `_private: ()` to `InlineOptions` should be the way to go, but it could be done separately, as such a change would only affect what the user could do and is not directly related to the new config option. 

The idea is that having this private field allows for fewer breaking changes. Right now, dependencies could build this struct directly with or without using `Default::default()` - a private field will make it impossible to build this struct directly, and the only possible way will be using a builder or `::default()`. For this reason, adding new config options will not break the dependencies

**Comment by kamilzych** [comment]
Ah, I see. Makes sense, although I'm not sure if it should be a part of this PR 🤔 

FYI, I've just committed a failing test, fixing it seems to be a bit harder than I thought and I have to run. I hope to find some time next week to work on it again. Thanks again! 🙏

**Comment by Stranger6667** [comment]
Sure thing! The private field thing is definitely for later!

Thank you again for working on this :) 
I'll take a deeper look over the weekend and hope that I will be able to finish things that are in the TODO list here :) Have a great weekend

**Comment by Stranger6667** [comment]
I think we are ready to go! :) I am so happy to see this feature implemented! :) Great work!

I will deal with Ruby separately

**Comment by kamilzych** [comment]
Awesome, glad to hear it! Perhaps I'll add more unit tests for corner cases (e.g. `at-rules` in external CSS, various flags' configs etc) in the future. Besides, once you release it, I'll test this feature performance-wise with real-world case (quite complex HTML with some at-rules I need to keep in output HTML).
Great collabo! 🎉

**Comment by Stranger6667** [comment]
Great! Also, feel free to add that HTML as a benchmark [here](https://github.com/Stranger6667/css-inline/blob/master/benchmarks/benchmarks.json) (if it could be shared publicly). The structure there could be extended so we can benchmark the new config option too (e.g. add optional `"keep_at_rules": true` to the new benchmark)
